packages/state-transition/src/cache/epochCache.ts

```diff
@@ -500,40 +576,79 @@ export class EpochCache {
 
   /**
    * Called to re-use information, such as the shuffling of the next epoch, after transitioning into a
-   * new epoch.
+   * new epoch. Also handles pre-computation of values that may change during the upcoming epoch and
+   * that get used in the following epoch transition.  Often those pre-computations are not used by the
+   * chain but are courtesy values that are served via the API for epoch look ahead of duties.
+   *
+   * Steps for afterProcessEpoch
+   * 1) update previous/current/next values of cached items
    */
   afterProcessEpoch(
-    state: BeaconStateAllForks,
+    state: CachedBeaconStateAllForks,
     epochTransitionCache: {
       nextEpochShufflingActiveValidatorIndices: ValidatorIndex[];
       nextEpochShufflingActiveIndicesLength: number;
       nextEpochTotalActiveBalanceByIncrement: number;
     }
   ): void {
-    this.previousShuffling = this.currentShuffling;
-    this.currentShuffling = this.nextShuffling;
-    const currEpoch = this.currentShuffling.epoch;
-    const nextEpoch = currEpoch + 1;
-
-    this.nextShuffling = computeEpochShuffling(
-      state,
-      epochTransitionCache.nextEpochShufflingActiveValidatorIndices,
-      epochTransitionCache.nextEpochShufflingActiveIndicesLength,
-      nextEpoch
-    );
+    const upcomingEpoch = this.nextEpoch;
+    const epochAfterUpcoming = upcomingEpoch + 1;
+    const nextActiveIndicesLength = epochTransitionCache.nextEpochShufflingActiveIndicesLength;
 
-    // Roll current proposers into previous proposers for metrics
+    // move current to previous
+    this.previousShuffling = this.currentShuffling;
+    this.previousDecisionRoot = this.currentDecisionRoot;
     this.proposersPrevEpoch = this.proposers;
 
-    const currentProposerSeed = getSeed(state, this.currentShuffling.epoch, DOMAIN_BEACON_PROPOSER);
-    this.proposers = computeProposers(currentProposerSeed, this.currentShuffling, this.effectiveBalanceIncrements);
+    // move next to current or calculate upcoming
+    this.currentDecisionRoot = this.nextDecisionRoot;
+    if (this.nextShuffling) {
+      // was already pulled by the api or another method on EpochCache
+      this.currentShuffling = this.nextShuffling;
+      this.shufflingCache?.metrics?.shufflingCache.nextShufflingOnEpochCache.inc();
+    } else {
+      this.currentShuffling =
+        this.shufflingCache?.getSync(upcomingEpoch, this.nextDecisionRoot, {
```

Comment:
should be tracked in a metric for this case. I expect most of the time `nextShuffling` should be available and it does not hit this whole "else" clause

Reply:
Updated.  Now that the nextShuffling is moved from the cache in the `build.then` it should basically always be available so I changed the metric to only show if its not already on the EpochCache.  And then the cache hit/miss is capture within the `shufflingCache.getSync` call.